internal/dag/dag.go

```diff
@@ -386,28 +402,45 @@ func (d *DAG) recompute() dag {
 		_d.roots = append(_d.roots, svh)
 	}
 
-	return _d
+	for meta, orph := range orphaned {
+		if orph {
+			ir, ok := d.ingressroutes[meta]
+			if ok {
+				status = append(status, ingressrouteStatus{object: ir, status: "orphaned", msg: "this IngressRoute is not part of a delegation chain from a root IngressRoute"})
+			}
+		}
+	}
+	return _d, status
 }
 
-func (d *DAG) processIngressRoute(ir *ingressroutev1.IngressRoute, prefixMatch string, visited map[meta]bool, host string, service func(m meta, port intstr.IntOrString) *Service, vhost func(host string, port int) *VirtualHost) {
-	// check if we have already visited this ingressroute. if we have, there is a cycle in the dag.
-	if visited[meta{name: ir.Name, namespace: ir.Namespace}] {
-		// TODO(abrand): Handle the cycle. Invalidate IngressRoute and set status?
-		return
-	}
+func (d *DAG) processIngressRoute(ir *ingressroutev1.IngressRoute, prefixMatch string, visited []*ingressroutev1.IngressRoute, host string, service func(m meta, port intstr.IntOrString) *Service, vhost func(host string, port int) *VirtualHost, orphaned map[meta]bool) []ingressrouteStatus {
+	visited = append(visited, ir)
+	defer func() {
```

Comment:
I don't think this code is doing what its supposed to, on exit from the function the callers value of visited will be unchanged. The memory beyond what the callers' copy of visited may contain `ir`, if there was sufficient space to append it without reallocating the slice, but the caller won't be able to see that.

Reply:
Ah definitely. I see how this would break down. Thanks @davecheney.

Do you have any suggestions here? My first thought is to build a stack type that has push/pop methods that we can use here, but wondering if there's a simpler approach.